pkg/agent/nodeportlocal/rules/iptable_rule.go

```diff
@@ -46,37 +47,57 @@ func NewIPTableRules() *iptablesRules {
 
 // Init initializes IPTABLES rules for NPL. Currently it deletes existing rules to ensure that no stale entries are present.
 func (ipt *iptablesRules) Init() error {
-	return ipt.CreateChains()
+	if err := ipt.InitRules(); err != nil {
+		return fmt.Errorf("initialization of NPL iptables rules failed: %v", err)
+	}
+	return nil
 }
 
-// CreateChains creates the chain NodePortLocalChain in NAT table.
-// All DNAT rules for NPL would be added in this chain.
-func (ipt *iptablesRules) CreateChains() error {
-	err := ipt.table.EnsureChain(iptables.NATTable, NodePortLocalChain)
-	if err != nil {
-		return fmt.Errorf("IPTABLES chain creation in NAT table failed for NPL with error: %v", err)
+// InitRules creates the NPL chain and link them to the PREROUTING (for incoming
+// traffic) and OUTPUT chain (for locally-generated traffic). All NPL DNAT rules
+// will be added to this chain.
+func (ipt *iptablesRules) InitRules() error {
+	if err := ipt.table.EnsureChain(iptables.NATTable, NodePortLocalChain); err != nil {
+		return err
 	}
 	ruleSpec := []string{
-		"-p", "tcp", "-j", NodePortLocalChain,
+		"-p", "tcp", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain,
 	}
-	err = ipt.table.EnsureRule(iptables.NATTable, iptables.PreRoutingChain, ruleSpec)
-	if err != nil {
-		return fmt.Errorf("IPTABLES rule creation in NAT table failed for NPL with error: %v", err)
+	if err := ipt.table.EnsureRule(iptables.NATTable, iptables.PreRoutingChain, ruleSpec); err != nil {
+		return err
+	}
+	if err := ipt.table.EnsureRule(iptables.NATTable, iptables.OutputChain, ruleSpec); err != nil {
+		return err
 	}
 	return nil
 }
 
+func buildRulesForPod(port int, podIP string) [][]string {
+	var rules [][]string
+	// For local traffic we need to masquerade the traffic to ensure that
+	// the gateway's IP is used as the source. Otherwise, traffic will be
+	// dropped by the SpoofGuard table in the OVS pipeline.
+	// Install this rule first as MARK is non-terminating but DNAT is.
+	rules = append(rules, []string{
+		"-p", "tcp", "-m", "tcp", "--dport", fmt.Sprint(port),
+		"-m", "addrtype", "--src-type", "LOCAL",
+		"-j", "MARK", "--or-mark", fmt.Sprintf("%#08x", types.NeedMasqueradeMark),
```

Comment:
If I'm wrong above, could it reuse the existing local output packet mark "0x1/0x1" to reduce the number of iptables rules (for both controlplane and dataplane's efficiency).
And it seems Kubernetes resets the mark before the packets is output:
```
-A KUBE-MARK-MASQ -j MARK --set-xmark 0x4000/0x4000
-A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
-A KUBE-POSTROUTING -j MARK --set-xmark 0x4000/0x0
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
```
Should we do the same? If we reuse 0x1, then it's intended to be kept.

Reply:
I can reset the mark. Not sure about re-using the `0x1` mark however, but maybe I am missing something. If I use the `0x1` mark, won't that automatically bypass NetworkPolicies? Isn't that the mark we use for liveness probes?

Reply:
Yes, it will bypass NetworkPolicies. But if the node accesses the local pod directly, it will bypass NetworkPolicies anyway. I'm not sure if this is a problem. We call the mark `HostLocalSourceMark`, not specific to liveness probe, though only used for it now.
```
	// HostLocalSourceBit is the bit of the iptables fwmark space to mark locally generated packets.
	// Value must be within the range [0, 31].
	HostLocalSourceBit = 0

	// HostLocalSourceMark is the mark generated from HostLocalSourceBit.
	HostLocalSourceMark = uint32(1 << HostLocalSourceBit)
```

Reply:
I see your point. I think I can reuse the mark then. Will update the PR soon.